### ISSUE TITLE
Can't convert ChunkedArray from a lazy array with a selection to a numpy array

### ISSUE THREAD

**Issue by beojan** [scikit-hep/uproot#323]
`np.array((tree['event']['m_hh'])[(200. < tree['event']['m_hh']) & (tree['event']['m_hh'] < 1000.)])`
gives the error:
`ValueError: could not broadcast input array from shape (2961863) into shape (9213873)`

**Comment by jpivarski** [comment]
Could you give the stack trace, so that I know what line number the error is on?

In small tests with my own examples, I wasn't able to reproduce a similar bug (including making the chunks in the lazy arrays different sizes from each other).

**Comment by beojan** [comment]
Slightly different specifics:
```
/usr/lib/python3.7/site-packages/awkward/array/chunked.py in __array__(self, *args, **kwargs)
    297                 out = self.numpy.empty(self.shape, dtype=self.dtype)
    298                 for chunk, slc in zip(self._chunks, self._slices()):
--> 299                     out[slc] = chunk
    300                 return out
    301         else:

ValueError: could not broadcast input array from shape (94144) into shape (106966)
```

**Comment by HDembinski** [comment]
I got a similar (the same?) issue.

```
import uproot
import numpy as np
import sys

data = uproot.lazyarrays(sys.argv[1:], "ana", ("vtx_len", "trk_len"))

# vtx_len = np.asarray(data["vtx_len"])
# trk_len = np.asarray(data["trk_len"])
vtx_len = data["vtx_len"]
trk_len = data["trk_len"]

counts = vtx_len[trk_len > 0]
w, xe = np.histogram(counts, bins=np.arange(20))  # this line is important
```
This raises a ValueError:
```
Traceback (most recent call last):
  File "plot_number_of_reconstructed_pvs.py", line 15, in <module>
    w, xe = np.histogram(counts, bins=np.arange(20))
  File "<__array_function__ internals>", line 6, in histogram
  File "/usr/local/lib/python3.7/site-packages/numpy/lib/histograms.py", line 783, in histogram
    a, weights = _ravel_and_check_weights(a, weights)
  File "/usr/local/lib/python3.7/site-packages/numpy/lib/histograms.py", line 278, in _ravel_and_check_weights
    a = np.asarray(a)
  File "/usr/local/lib/python3.7/site-packages/numpy/core/_asarray.py", line 85, in asarray
    return array(a, dtype, copy=False, order=order)
  File "/usr/local/lib/python3.7/site-packages/awkward/array/chunked.py", line 298, in __array__
    out[slc] = chunk
  File "/usr/local/lib/python3.7/site-packages/awkward/array/chunked.py", line 298, in __array__
    out[slc] = chunk
ValueError: could not broadcast input array from shape (371718) into shape (644127)
```

The same code works correctly when I use the commented lines instead.

**Comment by HDembinski** [comment]
I can try to attach a minimal tree here for you to reproduce this problem tomorrow.

**Comment by jpivarski** [comment]
That could be good, because as I see from the log, I wasn't able to reproduce this one.

By the way, sorry if I'm a little slow fixing things—I'm being snowed under by trying to do development while having these maintenance responsibilities.

**Comment by HDembinski** [comment]
# Full example to reproduce

[tree.zip](https://github.com/scikit-hep/uproot/files/3662578/tree.zip)
```
import uproot
import numpy as np
import sys

data = uproot.lazyarrays("tiny_sample.root", "ana", ("evt_bxtype", "vtx_len"))

# a = np.asarray(data["vtx_len"])
# b = np.asarray(data["evt_bxtype"])
a = data["vtx_len"]
b = data["evt_bxtype"]

counts = a[b > 0]
```
ufuncs don't seem to work on this masked array.
```
np.sum(counts)

Traceback (most recent call last):
  File "foo.py", line 14, in <module>
    np.sum(counts)
  File "<__array_function__ internals>", line 6, in sum
  File "/usr/local/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 2182, in sum
    initial=initial, where=where)
  File "/usr/local/lib/python3.7/site-packages/numpy/core/fromnumeric.py", line 88, in _wrapreduction
    return reduction(axis=axis, out=out, **passkwargs)
TypeError: sum() got an unexpected keyword argument 'axis'
```
```
np.histogram(counts)

Traceback (most recent call last):
  File "foo.py", line 14, in <module>
    np.histogram(counts)
  File "<__array_function__ internals>", line 6, in histogram
  File "/usr/local/lib/python3.7/site-packages/numpy/lib/histograms.py", line 783, in histogram
    a, weights = _ravel_and_check_weights(a, weights)
  File "/usr/local/lib/python3.7/site-packages/numpy/lib/histograms.py", line 278, in _ravel_and_check_weights
    a = np.asarray(a)
  File "/usr/local/lib/python3.7/site-packages/numpy/core/_asarray.py", line 85, in asarray
    return array(a, dtype, copy=False, order=order)
  File "/usr/local/lib/python3.7/site-packages/awkward/array/chunked.py", line 298, in __array__
    out[slc] = chunk
  File "/usr/local/lib/python3.7/site-packages/awkward/array/chunked.py", line 298, in __array__
    out[slc] = chunk
ValueError: could not broadcast input array from shape (223776) into shape (224302)
```

**Comment by jpivarski** [comment]
`np.sum(counts)` doesn't work because `np.sum` is not a ufunc. Awkward-array takes advantage of Numpy's [NEP 13](https://numpy.org/neps/nep-0013-ufunc-overrides.html) to override all ufuncs with `__array_ufunc__` (available since Numpy 1.13.1), but not [NEP 18](https://numpy.org/neps/nep-0018-array-function-protocol.html) to override all other Numpy functions, such as `np.sum`, with `__array_function__`.

Last I heard, NEP 18 was still being discussed, but now I see that it was released in Numpy 1.17 ([release notes](https://github.com/numpy/numpy/releases/tag/v1.17.0rc1), [milestone](https://github.com/numpy/numpy/issues/12028)). However, it's still in [provisional status](https://numpy.org/neps/index.html#provisional-neps-provisionally-accepted-interface-may-change), so the interface may change. I guess that means it's a TODO item for me to learn how to support it; however, a bigger project than this. It will probably go into awkward 1.0.

Even when I support NEP 18, I need to support Numpy versions < 1.17, so `np.sum(awkarray)` will fail for users. NEP 13 is absolutely essential (one gets wrong results, not an error message, if the ufunc calls `__array__` before computing the ufunc), so that's why the minimum Numpy version is 1.13.1.

By the way, awkward arrays have reducers:

```python
counts.sum()
```

and awkward 1.0 will have a suite of functions like

```python
awkward.sum(counts)
```

(similar to Dask, CuPy, and all the others waiting for NEP 18).

As for `np.histogram(counts)`, that works for me and it sounds familiar: I think it was a recently fixed bug in awkward-array. I'm using awkward 0.12.10.

**Comment by HDembinski** [comment]
I upgraded uproot but didn't upgrade awkward array. Dependencies are probably not automatically upgraded. Thank you for the very nice explanation.
